### DISCUSSION TITLE
Control Panel only mode/GraphQL Endpoint an admin route

### DISCUSSION THREAD

**Discussion by dwhoban** [craftcms/cms#12489] [Ideas] [answered]
I've been playing around with Headless mode and a number of different frameworks lately and keep coming back to the same point of configuration.

Using `baseCpUrl` + `headlessMode` gives a good result but there are a couple of scenarios that make things interesting.

AIUI based on the documentation you cannot access `graphql/api` directly and a route needs to be set.  This route maps to the url for each site.  This all works fine until you want to set `cpTrigger` to `null`.  Doing this means the API route returns a 404.

I found a work around for this with involved adding `define('CRAFT_CP', $_SERVER['REQUEST_URI'] !== '/api');` to `index.php` which works but has the potential to break things.

The request:

The goal of the feature is really to have a graphQL api route served on the `baseCpUrl` rather than the front end. 

This would be via a config flag which could just activate the route or it could be `cpOnlyMode` which does all the leg work and does something similar to `define(CRAFT_CP)`

Question:

In the interim, is there an alternative approach to handling the api endpoint via a controller or something else?

**Comment by brandonkelly** [comment, marked as answer]
You could register `api` as a custom CP route via the `EVENT_REGISTER_CP_URL_RULES` event from a custom module:

```php
use craft\events\RegisterUrlRulesEvent;
use craft\web\UrlManager;
use yii\base\Event;

Event::on(
    UrlManager::class,
    UrlManager::EVENT_REGISTER_CP_URL_RULES,
    function(RegisterUrlRulesEvent $event) {
        $event->rules['api'] = 'graphql/api';
    }
);
```

Alternatively, you could specify the controller action directly with an `action=graphql/api` query string param, rather than masking it with a route.
